cloudinit/sources/helpers/vmware/imc/config_nic.py

```diff
@@ -87,161 +80,190 @@ def find_devices(self):
-        @return (list): the subnets and routes list to configure the nic
+        Return the config dict needed to configure a nic
+        @return (dict): the config dict to configure the nic
         @param nic (NicBase): the nic to configure
         """
         mac = nic.mac.lower()
         name = self.mac2Name.get(mac)
         if not name:
             raise ValueError("No known device has MACADDR: %s" % nic.mac)
 
-        nics_cfg_list = []
-
-        cfg = {"type": "physical", "name": name, "mac_address": mac}
-
-        subnet_list = []
-        route_list = []
-
-        # Customize IPv4
-        (subnets, routes) = self.gen_ipv4(name, nic)
-        subnet_list.extend(subnets)
-        route_list.extend(routes)
-
-        # Customize IPv6
-        (subnets, routes) = self.gen_ipv6(name, nic)
-        subnet_list.extend(subnets)
-        route_list.extend(routes)
-
-        cfg.update({"subnets": subnet_list})
-
-        nics_cfg_list.append(cfg)
-        if route_list:
-            nics_cfg_list.extend(route_list)
-
-        return nics_cfg_list
-
-    def gen_ipv4(self, name, nic):
-        """
-        Return the set of subnets and routes needed to configure the
-        IPv4 settings of a nic
-        @return (set): the set of subnet and routes to configure the gateways
-        @param name (str): subnet and route list for the nic
-        @param nic (NicBase): the nic to configure
-        """
-
-        subnet = {}
-        route_list = []
-
+        nic_config_dict = {}
+        # match
+        match = self.gen_match(mac)
+        if match:
+            nic_config_dict.update(match)
+        # set-name
+        set_name = self.gen_set_name(name)
+        if set_name:
+            nic_config_dict.update(set_name)
+        # wakeonlan
+        wakeonlan = self.gen_wakeonlan(nic)
+        if wakeonlan:
+            nic_config_dict.update(wakeonlan)
+        # dhcp4
+        dhcp4 = self.gen_dhcp4(nic)
+        if dhcp4:
+            nic_config_dict.update(dhcp4)
+        # dhcp6
+        dhcp6 = self.gen_dhcp6(nic)
+        if dhcp6:
+            nic_config_dict.update(dhcp6)
+        # addresses
+        addresses = self.gen_addresses(nic)
+        if addresses:
+            nic_config_dict.update(addresses)
+        # routes
+        routes = self.gen_routes(nic)
+        if routes:
+            nic_config_dict.update(routes)
+        # nameservers
+        nameservers = self.gen_nameservers()
+        if nameservers:
+            nic_config_dict.update(nameservers)
+
+        return {name: nic_config_dict}
+
+    def gen_match(self, mac):
+        return {"match": {"macaddress": mac}}
+
+    def gen_set_name(self, name):
+        return {"set-name": name}
+
+    def gen_wakeonlan(self, nic):
         if nic.onboot:
-            subnet.update({"control": "auto"})
+            return {"wakeonlan": True}
+        else:
+            return {"wakeonlan": False}
 
+    def gen_dhcp4(self, nic):
```

Comment:
Nit: for the rest of these methods, do you think it makes more sense to return `{}` rather than `None`? The typing becomes more consistent and doesn't require `None` checks then.

Reply:
Yes, return `{}` now.